macros/trace_object/src/lib.rs

```diff
@@ -0,0 +1,196 @@
+            }
+        }
+    });
+
+    // Generate a fallback to the parent plan
+    let parent_field_delegator = if let Some(f) = parent_field {
+        let f_ident = f.ident.as_ref().unwrap();
+        let ref f_ty = f.ty;
+        quote! {
+            <#f_ty as PlanTraceObject #ty_generics>::trace_object::<T, KIND>(&self.#f_ident, __mmtk_trace, __mmtk_objref, __mmtk_worker)
+        }
+    } else {
+        quote! {
+            panic!("No more spaces to try")
+        }
+    };
+
+    quote! {
+        fn trace_object<T: crate::plan::TransitiveClosure, const KIND: crate::policy::gc_work::TraceKind>(&self, __mmtk_trace: &mut T, __mmtk_objref: crate::util::ObjectReference, __mmtk_worker: &mut crate::scheduler::GCWorker<VM>) -> crate::util::ObjectReference {
+            use crate::policy::space::Space;
+            use crate::policy::gc_work::PolicyTraceObject;
+            use crate::plan::transitive_closure::PlanTraceObject;
+            #(#space_field_handler)*
+            #parent_field_delegator
+        }
+    }
+}
+
+fn generate_create_scan_work<'a>(
+    space_fields: &[&'a Field],
+    scan_work_field: &Option<&'a Field>,
+    ty_generics: &TypeGenerics,
+) -> TokenStream2 {
+    if let Some(f) = scan_work_field {
+        // If the plan names a field for scan work, use it
+        let f_ident = f.ident.as_ref().unwrap();
+        let ref f_ty = f.ty;
+
+        quote! {
+            fn create_scan_work<E: crate::scheduler::gc_work::ProcessEdgesWork<VM = VM>>(&'static self, nodes: Vec<crate::util::ObjectReference>) -> Box<dyn crate::scheduler::GCWork<VM>> {
+                use crate::policy::gc_work::PolicyTraceObject;
+                <#f_ty as PolicyTraceObject #ty_generics>::create_scan_work::<E>(&self.#f_ident, nodes)
+            }
+        }
+    } else if !space_fields.is_empty() {
+        // If the plan does not name a specific field for scan work, just use the first space for scan work
+        let f = space_fields[0];
+        let f_ident = f.ident.as_ref().unwrap();
+        let ref f_ty = f.ty;
+
+        quote! {
+            fn create_scan_work<E: crate::scheduler::gc_work::ProcessEdgesWork<VM = VM>>(&'static self, nodes: Vec<crate::util::ObjectReference>) -> Box<dyn crate::scheduler::GCWork<VM>> {
+                use crate::policy::gc_work::PolicyTraceObject;
+                <#f_ty as PolicyTraceObject #ty_generics>::create_scan_work::<E>(&self.#f_ident, nodes)
+            }
+        }
+    } else {
+        // Otherwise, just panic
+        quote! {
+            fn create_scan_work<E: crate::scheduler::gc_work::ProcessEdgesWork<VM = VM>>(&'static self, nodes: Vec<crate::util::ObjectReference>) -> Box<dyn crate::scheduler::GCWork<VM>> {
+                panic!("Unable to create a scan work packet for the plan (the plan does not name a #[scan_work] field, or a #[trace] field")
+            }
+        }
+    }
+}
```

Comment:
Ditto. If https://github.com/mmtk/mmtk-core/issues/576 is applicable, this entire function can go away, too, like `PolicyTraceObject::create_scan_work` itself.